engine/consensus/approvals/assignment_collector.go

```diff
@@ -290,18 +266,22 @@ func (ac *AssignmentCollector) validateApproval(approval *flow.ResultApproval) e
 }
 
 func (ac *AssignmentCollector) ProcessApproval(approval *flow.ResultApproval) error {
+	// we have this approval cached already, no need to process it again
+	approvalCacheID := approval.Body.PartialID()
+	if cached := ac.verifiedApprovalsCache.Get(approvalCacheID); cached != nil {
+		return nil
+	}
+
```

Comment:
checks _first_ whether an approval is already cached and drops duplicates, _before_ doing expensive crypto verifications